### PR TITLE
Fixed #31842 -- Added DEFAULT_HASHING_ALGORITHM transitional setting.

### PR DESCRIPTION
**Pull request by felixxm**
ticket-31842

### CODE REVIEW

**Comment by apollo13** [comment]
From a security perspective I wonder if we should limit this that the existing `sha1` and the new `sha256` algos. We really don't ever want anyone to put `md5` there and it is really just a transitional setting.

File: django/conf/global_settings.py [hunk 1]
```diff
@@ -436,6 +436,11 @@ def gettext_noop(s):
 # you may be opening yourself up to a security risk.
 SECURE_PROXY_SSL_HEADER = None
 
+# Default hashing algorithm to use for encoding cookies, password reset tokens
+# in the admin site, user sessions, and signatures. Algorithm must be supported
+# by hashlib.
+DEFAULT_HASHING_ALGORITHM = 'sha256'
```
Comment (MarkusH) [inline django/conf/global_settings.py:442]: The default needs to be sha1 for existing projects. New projects can have the setting in the project template with `sha256`.

File: docs/ref/settings.txt [hunk 2]
```diff
@@ -1295,6 +1295,20 @@ Default email address to use for various automated correspondence from the
...
 and :setting:`MANAGERS`; for that, see :setting:`SERVER_EMAIL`.
 
+.. setting:: DEFAULT_HASHING_ALGORITHM
+
+``DEFAULT_HASHING_ALGORITHM``
+-----------------------------
+
+.. versionadded:: 3.1
+
+Default: ``sha256``
```
Comment (MarkusH) [inline docs/ref/settings.txt:1305]: Will need to be adjusted as well.

File: docs/releases/3.1.txt [hunk 3]
```diff
@@ -96,6 +96,15 @@ and generate and apply a database migration. For now, the old fields and
...
 
+``DEFAULT_HASHING_ALGORITHM`` settings
+--------------------------------------
+
+The new :setting:`DEFAULT_HASHING_ALGORITHM` setting allows defining the
+default hashing algorithm to use for encoding cookies, password reset tokens in
+the admin site, user sessions, and signatures created by
+:class:`django.core.signing.Signer` and :meth:`django.core.signing.dumps`.
+Algorithm must be supported by :py:mod:`hashlib`.
+
```
Comment (MarkusH) [inline docs/releases/3.1.txt:107]: Can we include a paragraph about how to upgrade? Something along these lines:

> When you're upgrading to Django 3.1 ensure that all your servers are running Django 3.1 before settings this value to `'sha256'` which is the recommended value. `'sha256'` will also be the default value from Django 4.0.

File: django/conf/global_settings.py (hunk 1 above)
Comment (apollo13) [reply django/conf/global_settings.py:442]: Does it? This is a specialized setting for people who need to run two versions in parallel. Those people should be able to read the release notes and prepare an upgrade plan. For everyone else (+new projects) `sha256` would be just fine as default.

File: docs/releases/3.1.txt (hunk 3 above)
Comment (apollo13) [reply docs/releases/3.1.txt:107]: See my comment above & I'll rather be "secure" by default. I don't want to push changing the default out to the year 2022.

**Comment by felixxm** [comment]
>     * There's @apollo13's query about whether we'll restrict the choice of available algorithm?

I was thinking about adding a system check for algorithms not supported by `hashlib`, so we could also check `md5`.

>     * Are we (immediately?) deprecating `DEFAULT_HASHING_ALGORITHM`?
> 
I think it's a nice feature per se (see [Simon's comment](https://groups.google.com/g/django-developers/c/D3c9--T-i-4/m/VeWmrUkPAQAJ)), so I would leave it.

We should also add to release notes a small explanation. (how to use it when upgrading multiple instances of the same app.)

**Comment by felixxm** [comment]
I added a system check.

**Comment by apollo13** [comment]
On Sun, Aug 2, 2020, at 13:54, Mariusz Felisiak wrote:
> I think it's a nice feature per se (see Simon's comment 
> <https://groups.google.com/g/django-developers/c/D3c9--T-i-4/m/VeWmrUkPAQAJ>), so I would leave it.

 We should make very clear that it doesn't support transparent upgrading then. Ie setting this to sha256, running with it and then changing it to sha512 will be a hard cut. There will be no migration path like there would be for sha1. Imo that limits the usefulness of such a setting for long term usage and makes it questionable if we want to support such usage.

File: docs/releases/3.1.txt (hunk 3 above)
Comment (MarkusH) [reply docs/releases/3.1.txt:117]: Thanks @felixxm. Looks good! :+1:

File: docs/releases/3.1.txt [hunk 4]
```diff
@@ -96,6 +96,25 @@ and generate and apply a database migration. For now, the old fields and
...
+--------------------------------------
+
+The new :setting:`DEFAULT_HASHING_ALGORITHM` transitional setting allows
+defining the default hashing algorithm to use for encoding cookies, password
+reset tokens in the admin site, user sessions, and signatures created by
+:class:`django.core.signing.Signer` and :meth:`django.core.signing.dumps`.
+
+If you upgrade multiple instances of the same app to Django 3.1, you should
+set :setting:`DEFAULT_HASHING_ALGORITHM` to ``'sha1'``, because hashes that use
+SHA-1 algorithm are valid in Django 3.1 and in older versions of Django. You
```
Comment (carltongibson) [inline docs/releases/3.1.txt:111]: Just this one line that I'm not sure on.

Pre-coffee sketch:

> Support for SHA-256 was added in Django 3.1. If you are uprading multiple instances of the same app to Django 3.1, you should set ... to `'sha1'` during the transition, in order to allow compability with the older versions of Django. Once the transition to 3.1 is complete...

Comment (pope1ni) [reply docs/releases/3.1.txt:111]: Should it be "project" rather than "app"?

Comment (carltongibson) [reply docs/releases/3.1.txt:111]: Probably "pre-coffee"

Comment (pope1ni) [reply docs/releases/3.1.txt:111]: It's not you, it was in the original too… 😉☕

Comment (felixxm) [reply docs/releases/3.1.txt:111]: Updated.

File: docs/releases/3.1.txt [hunk 5]
```diff
@@ -96,6 +96,27 @@ and generate and apply a database migration. For now, the old fields and
 transforms are left as a reference to the new ones and are :ref:`deprecated as
 of this release <deprecated-jsonfield>`.
 
+.. _default-hashing-algorithm-usage:
+
+``DEFAULT_HASHING_ALGORITHM`` settings
+--------------------------------------
+
+The new :setting:`DEFAULT_HASHING_ALGORITHM` transitional setting allows
+defining the default hashing algorithm to use for encoding cookies, password
```
Comment (smithdc1) [inline docs/releases/3.1.txt:105]: At the risk of being a pedant, I keep tripping over the word 'defining' here. Here are a couple of suggestions:
'... defining of the...'
'... transitional setting specifies the...'

Comment (felixxm) [reply docs/releases/3.1.txt]: OK, I changed to "specifying".

**Comment by felixxm** [comment]
Merged in d907371ef99a1e4ca6bc1660f57d81f265750984. Thanks y'all :confetti_ball:

**Pull request closed without merging**